File: ruby/binary-search/binary_search.rb
# Performs a Binary Search for a particular int in an array
class BinarySearch
  attr_reader :list
  def initialize(list)
    @list = list
    fail ArgumentError if list != list.sort
  end

  def search_for(val)
    search(val, 0, list.size - 1)
  end

  def middle(left = 0, right = list.size - 1)
    (left + right) / 2
  end

  private

  def search(val, left, right)
    fail RuntimeError if left >= right

    mid = middle(left, right)
    return search(val, left, mid - 1) if val < list[mid]
    return search(val, mid + 1, right) if val > list[mid]
    mid
  end
end

File: python/binary-search/binary_search.py
def binary_search(array, num_to_find):
    return binary_search_with_bounds(array, num_to_find, 0, len(array) - 1)

def binary_search_with_bounds(array, num_to_find, left, right):
    if (left <= right):
        middle = left + (right - left) // 2
        middle_value = array[middle]

        if (num_to_find == middle_value):
            return middle
        elif (num_to_find < middle_value):
            return binary_search_with_bounds(array, num_to_find, left, middle - 1)
        elif (num_to_find > middle_value):
            return binary_search_with_bounds(array, num_to_find, middle + 1, right)
    else:
        raise ValueError("Value not found")
